### PR TITLE
Crank nickelson#3

### PR DESCRIPTION
**Pull request by TomHrm**

### CODE REVIEW

File: Backend/Solver/CrankNicolson/Iteration.c [hunk 1]
```diff
@@ -1 +1,93 @@
-// TODO: Iterate
\ No newline at end of file
+#include <stdio.h>
```
Comment (MetaColon) [inline Backend/Solver/CrankNicolson/Iteration.c:1]: This file should be deleted

File: Backend/Solver/CrankNicolson/Iteration.h [hunk 2]
```diff
@@ -0,0 +1,9 @@
+#ifndef BACKEND_NICKELSON_H
```
Comment (MetaColon) [inline Backend/Solver/CrankNicolson/Iteration.h:1]: This file should also be deleted.

File: Backend/Solver/Matrix/Utility.c [hunk 3]
```diff
@@ -4,123 +4,219 @@
...
+        exit(0);
+    }
+    int r = a.rowCount;
+    int c = a.columnCount;
+    Matrix result = createMatrix(r,c);
+    int i,j;
+    for(i = 0; i < r ; i++){
+        for(j = 0; j < c; j++){
+            result.matrix[i][j] = a.matrix[i][j] + b.matrix[i][j];
+            //result[i][j] = a[i][j]+b[i][j]
```
Comment (MetaColon) [inline Backend/Solver/Matrix/Utility.c:114]: We don't need commented code.

File: Backend/Solver/Matrix/Utility.c [hunk 4]
```diff
@@ -4,123 +4,219 @@
...
+        exit(0);
+    }
+    int r = a.rowCount;
+    int c = a.columnCount;
+    Matrix result = createMatrix(r,c);
+    int i,j;
+    for(i = 0; i < r ; i++){
+        for(j = 0; j < c; j++){
+            result.matrix[i][j] = a.matrix[i][j] - b.matrix[i][j];
+            //result[i][j] = a[i][j]-b[i][j]
```
Comment (MetaColon) [inline Backend/Solver/Matrix/Utility.c:134]: See above

File: Backend/Solver/Matrix/Utility.c [hunk 5]
```diff
@@ -4,123 +4,219 @@
...
+                result.matrix[i][k] += a.matrix[i][j] * b.matrix[j][k];
+        }
+    return result;
+}
 
-    int n = a->rowSize;
+Matrix sum(Matrix a, Matrix b)
+{
+    if(a.rowCount != b.rowCount || a.columnCount != b.columnCount){
+        fprintf(stderr, "Error: columns of Matrix a != rows of Matrix b");
```
Comment (MetaColon) [inline Backend/Solver/Matrix/Utility.c:104]: Wrong error message

File: Backend/Solver/Matrix/Utility.c [hunk 6]
```diff
@@ -4,123 +4,219 @@
...
+        exit(0);
```
Comment (MetaColon) [inline Backend/Solver/Matrix/Utility.c:105]: 0 is not an error exit code

File: Backend/Solver/Matrix/Utility.c [hunk 7]
```diff
@@ -4,123 +4,219 @@
...
+        }
+    }
+
+    return result;
+}
+
+Matrix subtract(Matrix a, Matrix b)
+{
+    if(a.rowCount != b.rowCount || a.columnCount != b.columnCount){
+        fprintf(stderr, "Error: columns of Matrix a != rows of Matrix b");
```
Comment (MetaColon) [inline Backend/Solver/Matrix/Utility.c:124]: See above

File: Backend/Solver/Matrix/Utility.c [hunk 8]
```diff
@@ -4,123 +4,219 @@
...
-                result.matrix[i][j] += a->matrix[i][k] * b->matrix[k][j];
+            result.matrix[i][j] = a.matrix[i][j] * f;
         }
     }
     return result;
 }
 
-// n is the dimension of a->matrix
-double determinant(Matrix *a)
+// n is the dimension of a.matrix
```
Comment (MetaColon) [inline Backend/Solver/Matrix/Utility.c:176]: This comment doesn't seem valid anymore

File: Backend/Tests/Check_Matrix.h [hunk 9]
```diff
@@ -1,172 +1,332 @@
...
+    freeMatrix(result);
+}
+END_TEST
+
+START_TEST(test_factor)
+{
+    printf("\n");
+    printf("test_factor \n");
+    Matrix input = createMatrix(2, 2);
+    input.matrix[0][0] = 2.0;
```
Comment (MetaColon) [inline Backend/Tests/Check_Matrix.h:214]: The matrices for the tests shouldn't have the same value in all columns. E.g.

    1 2
    3 4
would be a better testcase, as it ensures the correct ordering is preserved.

This is not only relevant for `test_factor`, but also for the other tests.

File: Backend/Tests/Check_Matrix.h [hunk 10]
```diff
@@ -1,172 +1,332 @@
...
+    Matrix input_2 = createMatrix(5,1);
+    for(int i = 0; i < 5; i++)
+    {
+        input_2.matrix[i][0] = -1;
+    }
+    printf("Matrix 2: \n");
+    printMatrix(input_2);
+    Matrix thomasMatrix = thomasSolve(input_1, input_2);
+    printMatrix(thomasMatrix);
+    printf("results should: \n");
```
Comment (MetaColon) [inline Backend/Tests/Check_Matrix.h:320]: There is no validation happening here

**Review by MetaColon — CHANGES_REQUESTED** [review]
Just a few small changes required. Please note that, as I did a lot of edits in this branch too, you should review my changes as well.

File: Backend/Solver/Matrix/Utility.c [hunk 11]
```diff
@@ -122,7 +121,7 @@ Matrix subtract(Matrix a, Matrix b)
 {
     if(a.rowCount != b.rowCount || a.columnCount != b.columnCount){
         fprintf(stderr, "Error: columns of Matrix a != rows of Matrix b");
```
Comment (MetaColon) [inline Backend/Solver/Matrix/Utility.c:123]: This error message is still wrong

**Review by MetaColon — CHANGES_REQUESTED** [review]
Just one adjustment needed.

File: Backend/Program.c [hunk 12]
```diff
@@ -2,34 +2,48 @@
...
 
 int main(int argc, char *argv[])
 {
-    double value;
-    Function function;
+    Function potential, initial;
     if (argc != 3)
     {
-        fprintf(stderr, "Expected %s <function> <value> (2 arguments), but got %d argument%s\n",
+        fprintf(stderr, "Expected %s <potential function> <initial function> (2 arguments), but got %d argument%s\n",
```
Comment (TomHrm) [inline Backend/Program.c:13]: Is Github displaying the Format wrong here?

Comment (MetaColon) [reply Backend/Program.c:13]: Looks alright to me in GitHub.

**Review by MetaColon — APPROVED** [review]

**Pull request merged**